### DISCUSSION TITLE
RPi Pico/CircuitPython MIDI Device not recognised

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#197] [Q&A]
I hooked up a Raspberry Pi Pico running CircuitPython as a USB MIDI device and it didn't seem to get recognised.  Attached is a printout from the display on plugging it in.

I'm not very familiar with USB devices/etc, but when I used it with MidiOx on my PC it appears as "CircuitPython Audio".  I believe the relevant descriptors, etc, are as described here: https://github.com/adafruit/circuitpython/blob/main/shared-module/usb_midi/__init__.c

Any thoughts?

Thanks,
Kevin

![IMG_6162](https://user-images.githubusercontent.com/68612569/166305199-d59899a6-1069-47bb-b6a2-2a3283286934.JPG)

**Comment by probonopd** [comment]
Does it work with this firmware? https://github.com/rsta2/pico/tree/main/midi_adapter

Compiled version: https://github.com/probonopd/MiniDexed/files/8240422/midi_adapter.uf2.zip

**Comment by probonopd** [comment]
I also had it working with Raspberry Pi Pico running CircuitPython as a USB MIDI device at some point. Do you have the CircuitPython MIDI libraries and their dependencies installed on the Pico?

**Reply by diyelectromusic** [reply]
> I also had it working with Raspberry Pi Pico running CircuitPython as a USB MIDI device at some point. Do you have the CircuitPython MIDI libraries and their dependencies installed on the Pico?

Yes.  It works fine with a PC.  It is just using the same Adafruit USB MIDI libraries I've used many times before for other projects.  I ran up the firmware with USB MIDI logging enabled and can see notes from my USB keyboard controller, but nothing from the Pico.

That other firmware is recognised fine.

**Reply by probonopd** [reply]
If you want to see MIDI coming from a keyboard attached to MiniDexed on your Pico, then you need to configure Midi THRU in the `minidexed.ini` file accordingly.

**Reply by diyelectromusic** [reply]
> If you want to see MIDI coming from a keyboard attached to MiniDexed on your Pico, then you need to configure Midi THRU in the `minidexed.ini` file accordingly.

No, its the other way round - I'm trying to get the Pico to act as a simple button controller (in addition to a keyboard) for the MiniDexed, but as you can see from the log trace above, it just isn't being recgonised as a USB MIDI device.

**Reply by probonopd** [reply]
These messages don't necessarily need to mean that. I think CircuitPython simulates many different USB endpoints, not all of which MiniDexed can use. In any case, I had success with using CircuitPython to play notes on MiniDexed using a RPi Pico.

**Reply by diyelectromusic** [reply]
Ok, thanks.  I'll do some more experimenting and see if I can pin it down!  I haven't ruled out "user error, replace user" yet... Python and I don't always get on :)

**Reply by probonopd** [reply]
Maybe you are missing some needed Python libraries?

**Reply by diyelectromusic** [reply]
> Maybe you are missing some needed Python libraries?

As I say, a PC can see it fine - I get the correct MIDI messages in MIDIOx, so I know the Pico is doing some correct USB MIDI things.  I'll have another look later.

**Reply by probonopd** [reply]
Wrong wiring? TX needs to go to RX, and RX to TX, and GND must be connected to GND.

**Reply by diyelectromusic** [reply]
Ok, here is the latest.  I have three RPi Pico "keyboards" I'm trying, all running CircuitPython using the Adafruit USB MIDI library and at the same time, the UART for an optional serial MIDI link.  Two are "well established" toy keyboards(1) that do both serial MIDI and USB MIDI, the third is the thing I've knocked together for MiniDexed just recently.

I have a V1 Pi and a V4 Pi MiniDexed, and a V1 RPi MiniSynth from Rene Stange using his Circle environment (2). I also have a PC running MIDI Ox with either direct USB MIDI or serial MIDI via my Roland UM-ONE. I also have a V4 Pi running Zynthian(3) and a V3 Pi running MT32-Pi(4).

I also have a xKeys USB MIDI controller.  Here are the results:

xKeys USB: MiniDexed - works; miniSynth - works; MT32-Pi - works; Zynthian - works; PC - works.
Pico in serial MIDI mode: MiniDexed - works; miniSynth - works; MT32-Pi - works; Zynthian - works.  All using Clumsy MIDI interface for serial MIDI IN.  PC - works too via Roland UM-ONE USB/MIDI interface.  Same for all three Picos.
Pico in USB MIDI mode: MiniDexed - **doesn't work**; miniSynth - works; MT32-Pi - works; Zynthian - works; PC - works.  Same for all three Picos.

With MIDI debugging going on I can see note messages in all "working" cases with miniDexed and get nothing when it isn't working.  In fact I can plug in my Pico USB, see the errors as indicated above but have nothing happen and then attach the serial MIDI and it starts working and I can see Note On/Off messages.

So, if I'm doing something wrong (which I still haven't ruled out) then I've done it wrong three times, but in a way that the Circle-based miniSynth and MT32-Pi still both work, and Raspian/Zynthian and a Windows PC are all fine with too...

By the way, everything is configured for MIDI channel 1, although I've tried OMNI too.

Any further thoughts? I'm running out of ideas :)

Thanks,
Kevin

(1) Details here: https://diyelectromusic.wordpress.com/2022/04/08/yet-another-toy-keyboard-usb-midi-controller-part-2/ - where I'm using it with Zynthian as USB MIDI and here: https://diyelectromusic.wordpress.com/2021/06/28/toy-keyboard-usb-matrix-decode-with-the-pi-pico-part-2/ where I've used it as a test keyboard for a range of projects since June last year.
(2) https://diyelectromusic.wordpress.com/2022/03/20/bare-metal-raspberry-pi-synth-part-4/
(3) https://diyelectromusic.wordpress.com/2021/03/08/zynthian-and-clumsy-midi/
(4) https://diyelectromusic.wordpress.com/2021/03/01/raspberry-pi-clumsy-midi-and-mt32-pi/

**Reply by diyelectromusic** [reply]
> Wrong wiring? TX needs to go to RX, and RX to TX, and GND must be connected to GND.

On the Pi or Pico's built-in USB ports!? :)  As I say, serial MIDI is fine...

**Reply by probonopd** [reply]
I was referring to serial MIDI.

If you are using USB MIDI then it should just work. 
This works for me:

```
import time
import random
import usb_midi
import adafruit_midi

from adafruit_midi.note_on import NoteOn
from adafruit_midi.note_off import NoteOff
from adafruit_midi.pitch_bend import PitchBend
from adafruit_midi.control_change import ControlChange

midi = adafruit_midi.MIDI(midi_out=usb_midi.ports[1], out_channel=0)

print("Midi test")

print("Default output MIDI channel:", midi.out_channel + 1)

i = 0
note = 44

while i < 500:
    midi.send(NoteOn(note, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+1, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+2, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+3, 120))
    time.sleep(0.00)    
    midi.send(NoteOn(note+4, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+5, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+6, 120))
    time.sleep(0.00)
    midi.send(NoteOn(note+7, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note, 120))
    time.sleep(0.00)
    
    time.sleep(0.2)
    
    midi.send(NoteOff(note+1, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note+2, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note+3, 120))
    time.sleep(0.00)    
    midi.send(NoteOff(note+4, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note+5, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note+6, 120))
    time.sleep(0.00)
    midi.send(NoteOff(note+7, 120))
    time.sleep(0.00)
    
    time.sleep(0.1)

    i = i+1
    note = note + 1
    if note > 88:
        note = 44
```

In the lib folder, I have a file `adafruit_pioasm.py` and a folder `adafruit_midi`.

`boot_out.txt` says:

```
Adafruit CircuitPython 7.2.0 on 2022-02-24; Raspberry Pi Pico with rp2040
Board ID:raspberry_pi_pico
```

**Reply by diyelectromusic** [reply]
I think I might have a breakthrough!  It works on the MiniDexed Pi V1 but not on the Pi V4. I'm now wondering if it is related to the type of USB subsystem in the hardware given this clue: https://github.com/rsta2/circle/blob/bbc9264ead9dc9e0cf5fc2d4530c26b55ca78c29/include/circle/usb/usbhcidevice.h

As it works with my V3 MT32-Pi too, that might back this up?  What type of Pi are you using?

So this might be something to do with the upstream support for USB on the Pi 4 in circle...  If I get a chance, I'll try MiniDexed on my Pi V3...

**Reply by diyelectromusic** [reply]
Yep, works fine on a Pi V3, so must be something weird in the Pi 4 USB stack on circle I guess?

Update: Ok, not quite as simple as a problem with Circle though, as it works fine on both a V3 and V4 MT32-Pi... so it is some odd combination of MiniDexed, the version of Circle being used, and a Pi V4...

**Reply by diyelectromusic** [reply]
Looking at that "xhciep" log entry, I'm guessing that is coming from here: https://github.com/rsta2/circle/blob/master/lib/usb/xhciendpoint.cpp#L226

The question is why... I don't really know enough about USB stacks to go much further...

(Note: I thought that log entry looked like xhcicp, but that doesn't seem to be a legit log subsystem in Circle, so I'm guessing it is xhciep)

**Reply by diyelectromusic** [reply]
I've now reproduced this with the built-in Circle miniorgan sample app, so I've logged this as an issue against Circle: https://github.com/rsta2/circle/issues/293

**Reply by probonopd** [reply]
Thanks for testing @diyelectromusic

**Reply by diyelectromusic** [reply]
So it turns out the issue was related to the int3-0-0 failure further up.  Adding the line usbignore=int3-0-0 into a cmdline.txt file on the SD card makes Circle ignore that interface and then the following initialisation of int1-3-0 (the audio/MIDIStreaming interface) works fine! I'll see if I can make a note somewhere on the wiki...

Notes added to: https://github.com/probonopd/MiniDexed/wiki/Files#minidexedini

**Reply by diyelectromusic** [reply]
Another solution is to disable the HID interface on the CircuitPython side.  To do this, create a boot.py file that contains the following:
```
import usb_hid, usb_midi
usb_hid.disable()
usb_midi.enable()
```
Full details of CircuitPython USB device customisation can be found here: https://learn.adafruit.com/customizing-usb-devices-in-circuitpython?view=all

**Reply by probonopd** [reply]
Good to know!
